tcclient/creds.go

```diff
@@ -7,16 +7,75 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"net/http"
 	"strconv"
 	"strings"
 	"time"
 
 	"github.com/taskcluster/jsonschema2go/text"
 	"github.com/taskcluster/slugid-go/slugid"
+	"github.com/tent/hawk-go"
 )
 
-type Credentials struct {
-	ClientId    string `json:"clientId"`
+type Credentials interface {
+	fmt.Stringer
+	ConfigureAuth(httpRequest *http.Request) error
+	getExtField() (string, error)
+	HawkCredentials() *hawk.Credentials
```

Comment:
Either the interface exposes what you need to call Hawk, or it takes URLs and requests and calls hawk for you. So either do:
```go
type Credentials interface {
  fmt.Stringer
  HawkExt() string
  HawkCredentials() *hawk.Credentials
}
```
or you do: (this is better)
```go
type Credentials interface {
  fmt.Stringer
  SignRequest(req *http.Request) error
  SignURL(url string, duration time.Duration) url.URL, error
}
```


----------
I'm also okay with:
```go
type Credentials interface {
  fmt.Stringer
  SignRequest(req *http.Request) error
  SignURL(url *url.URL, duration time.Duration) error
}
```
But really we can probably do:
```go
type Credentials interface {
  fmt.Stringer
  SignRequest(req *http.Request)
  SignURL(url *url.URL, duration time.Duration)
}
```
If we are a tiny bit smart we can check for anything that would cause an error when we create the `Credentials` object... Thus no error will happen during `SignUrl` or `SignRequest`, if an error happens there we just panic on it...

Hmm, maybe getting passed nil as `req` or `url` is bad, but we could just choose to ignore that silently.
@petemoore Now that credentials are immutable is there really any error we can't check for upfront?

I think this would remove a lot of error checking and pain down the line :) yay for fewer errors to handle.